### PR TITLE
Fix - NKAI not using monoliths and whirlpools

### PR DESCRIPTION
**Pull request by MichalZr6**
Probably fixes bugs: #3599 #4216 #4104

### CODE REVIEW

**Comment by MichalZr6** [comment]
Sorry, I might have done this commit too soon... Have to test it.

**Comment by nullkiller** [comment]
Actually this logic is for a different branch. Monolith itself is useless for AI. oneways are blocked at pathfinder as we do not know if we can get out of it. But I can suggest easy hack: hero is main and army strength like 20 000 or more. This one is a bit more reliable but still AI can get into a trap. The most reliable is to try figure out if a hero can get back from there but in case of FoW it can not be accomplished anyway and complex to implement.
Wirlpools, are you sure they do not work? Might be because they cause army losses. Btw theoretically this can be reflected by pathfinder. It has own column for army loss so adding there will allow tracking this fact and make decisions accordingly.

**Comment by nullkiller** [comment]
AIPathfinderConfig is responsible for selecting allowed teleports. But it is still tricky as additional changes are required

**Comment by nullkiller** [comment]
If you feel like enabling them is useful you can introduce configuration settings instead and mod. One day I will fix it but not now.

**Comment by DjWarmonger** [comment]
IIRC my logic in VCAI was to use Monoliths for exploration. Just by entering it you reveal part of the map. Plus, this is typically done by scouts, so risk is lower.

**Comment by nullkiller** [comment]
That may be ported to NKAI. At least with twoway it works exactly like this.

File: AI/Nullkiller/AIUtility.cpp [hunk 1]
```diff
@@ -393,11 +393,12 @@ bool shouldVisit(const Nullkiller * ai, const CGHeroInstance * h, const CGObject
 		}
 		return false;
 	}
-	case Obj::MONOLITH_ONE_WAY_ENTRANCE:
 	case Obj::MONOLITH_ONE_WAY_EXIT:
+		return false;
+	case Obj::MONOLITH_ONE_WAY_ENTRANCE:
 	case Obj::MONOLITH_TWO_WAY:
 	case Obj::WHIRLPOOL:
-		return false;
```
Comment (nullkiller) [inline AI/Nullkiller/AIUtility.cpp:400]: This should be left as is. The mechanics of entering one-way monolith is a bit different and should relay on pathfinder instead. You can check my PR regarding whirlpools #4313 .

**Pull request closed without merging**